ts/deckoptions/ConfigSelector.svelte

```diff
@@ -52,6 +53,6 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
             {/each}
         </select>
 
-        <OptionsDropdown {state} />
+        <svelte:component this={optionsDropdown.component} {...optionsDropdown} />
```

Comment:
The code reuse is a definite improvement, but it's unfortunate that we need to use dynamic components here - I feel like they make the code harder to follow, as we're splitting the UI logic up into a separate svelte and js file, and having to store things in a temporary variable. The dynamic components are necessary to provide an API for add-ons, but for our own usage inside Svelte files, perhaps it would be nicer if we could reference the underlying component classes directly, instead of having to move logic into separate js files?

At the moment ButtonsGroup and DropdownMenu rely on a prop to pass in a list of dynamic components, which makes them difficult to reference directly in a Svelte file. If we changed them to something like this:

DropdownMenu:
```html
<ul {id} class="dropdown-menu">
    <slot />
</ul>
```

DropdownMenuItem:
```html
<li>[existing html]</li>
```

Then I think we could change a reactive assignment from

```js
$: menu = dropdownMenu({
                items: [
                    dropdownItem({
                        label: "Add Config",
                        onClick: addConfig,
                    }),
                    dropdownItem({
                        label2: "Rename Config",
                        onClick: renameConfig,
                    }),
                    dropdownItem({
                        label: "Remove Config",
                        onClick: removeConfig,
                    }),
                    dropdownDivider({}),
                    dropdownItem({
                        label: "Save to All Children",
                        onClick: () => save(true),
                    }),
                ],
```

to something like this in the Svelte body:

```html
<DropdownMenu>
    <DropdownMenuItem label="Add Config" on:click={addConfig} />
    <DropdownMenuItem label="Rename Config" on:click={renameConfig} />
    <DropdownMenuItem label=Remove Config" on:click={removeConfig} />
    <DropdownDivider />
    <DropdownItem label="Save to All Children" on:click={() => save(true)} />
</DropdownMenu>
```

For ButtonGroup, since we can't iterate over slots, I guess we would need to teach each kind of button how to hide itself instead, which is more verbose, but a bit less magic. And then each button could be passed in like is done with the dropdown menu above.

I feel like that makes the code a fair bit easier to follow. On the other hand, I have not actually implemented this, and there may be something I've missed. What do you think?

Sadly it looks like Svelte [doesn't currently provide an API](https://github.com/sveltejs/svelte/pull/5687) for setting slots from within JS, so if we were to go ahead with this change, to support the JS API as well without code duplication, we'd probably need to make a wrapper component that takes the dynamic components as props, and embeds them in a slot.

Reply:
I also have to agree that slots would certainly _feel_ better, but it also feels like we're settling with a strictly less powerful version of what we can do with dynamic components over better looking syntax. If we were to build up a "GUI component library", should we have two identical versions, one using slots, and one using dynamic components? The two things you already mentioned, and which make me very wary of slots, is that you 1. iterate over slots, and 2. you cannot set them from outside Svelte files. So whenever we'll use slots, it's like we're making it harder for ourselves.

Reply:
I do have _some_ experience with React, so I'm somewhat aware, that a lot of the things I'm arguing for is basically how you'd do it in React, like passing handlers (`onClick`) directly into the component, or passing components as props, rather than slotting them. It's not like I do it on purpose, it's more like, in striving to have more versatile components, it's just what seems most natural 🤔